### PR TITLE
refactor(compiler): drop regular imports when symbols can be defer-loaded

### PR DESCRIPTION
**Pull request by AndrewKushnir**
This commit updates the logic to drop regular imports when all symbols that it brings can be defer-loaded.
The change ensures that there is no mix of regular and dynamic imports present in a source file.

## PR Type
What kind of change does this PR introduce?

- [x] Refactoring (no functional changes, no api changes)

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/compiler-cli/test/ngtsc/ngtsc_spec.ts [hunk 1]
```diff
@@ -8872,23 +8872,219 @@ function allTests(os: string) {
       });
     });
 
-    // TODO: replace with tests for `defer`.
-    // TODO: maybe put deferred tests in a separate file?
     describe('deferred blocks', () => {
-      it('should not error for deferred blocks', () => {
+      it('should handle deferred blocks', () => {
         env.tsconfig({_enabledBlockTypes: ['defer']});
+        env.write('cmp-a.ts', `
```
Comment (kbrilla) [inline packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:8878]: not sure if already covered but want to make sure
What about imports like 
```
import {CmpA, type TypeOnly} from 'a';
//CmpA used in defer block but TypeOnly in ts
```
or
```
import {CmpA} from 'a';
import type {TypeOnly} from 'a';
//CmpA used in defer block but TypeOnly in ts so dynamic import can be used I think as type import will be droped
```
or
```
import {CmpA} from 'a';
import {type TypeOnly} from 'a';
//CmpA used in defer block but TypeOnly in ts but import {type TypeOnly} from 'a'; will be  import {} from 'a' so no dynamic import can be used
```
or
```
import {CmpA} from 'a'; 
import {CmpB} from 'a';
//both only used in defer so dynamic import can be used 
``` 

```
import {CmpA} from 'a'; 
import {CmpB} from 'a';
//CmpA used in defer but egearly so no dynamic import can occur
``` 

not sure if https://www.typescriptlang.org/tsconfig#importsNotUsedAsValues and https://www.typescriptlang.org/tsconfig#verbatimModuleSyntax will influence this

Comment (AndrewKushnir) [reply packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:8878]: @kbrilla thanks for the comment! We plan to expand the coverage of different import shapes, this PR just covers the most common ones.

**Comment by AndrewKushnir** [comment]
[Exploratory presubmit](https://test.corp.google.com/OCL:553034331:BASE:553034227:1690955772876:b9dc73a2).

File: packages/compiler-cli/src/ngtsc/annotations/common/src/util.ts [hunk 2]
```diff
@@ -331,7 +331,8 @@ export function createSourceSpan(node: ts.Node): ParseSourceSpan {
  */
 export function compileResults(
     fac: CompileResult, def: R3CompiledExpression, metadataStmt: Statement|null, propName: string,
-    additionalFields: CompileResult[]|null): CompileResult[] {
+    additionalFields: CompileResult[]|null,
+    deferrableImports?: Set<ts.ImportDeclaration>): CompileResult[] {
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/annotations/common/src/util.ts:335]: In the compiler we typically avoid optional arguments - this should be `|null` instead.

File: packages/compiler-cli/src/ngtsc/transform/src/api.ts [hunk 3]
```diff
@@ -248,6 +248,7 @@ export interface CompileResult {
   initializer: Expression|null;
   statements: Statement[];
   type: Type;
+  deferrableImports?: Set<ts.ImportDeclaration>;
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/transform/src/api.ts:251]: Same here - avoid optional fields.

**Review by alxhub — APPROVED** [review]

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:553034331:BASE:553554069:1691090381566:a27b70b2).

**Pull request closed without merging**